[SPARK-22170][SQL] Reduce memory consumption in broadcast joins.

rdblue opened the pull request:
## What changes were proposed in this pull request?

This updates the broadcast join code path to lazily decompress pages and
iterate through UnsafeRows to prevent all rows from being held in memory
while the broadcast table is being built.

## How was this patch tested?

Existing tests.

rdblue committed:
SPARK-22170: Reduce memory consumption in broadcast joins.

This updates the broadcast join code path to lazily decompress pages and
iterate through UnsafeRows to prevent all rows from being held in memory
while the broadcast table is being built.

SparkQA commented:
**[Test build #82338 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82338/testReport)** for PR 19394 at commit [`d82ffca`](https://github.com/apache/spark/commit/d82ffca19c84589dfcb87a81996fb3ce2e31e91a).

rdblue commented:
Ideally, this would also use a TaskMemoryManager so the driver can spill results to disk instead of dying with an OOM. Is there any plan to add a memory manager for the driver?

SparkQA commented:
**[Test build #82338 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82338/testReport)** for PR 19394 at commit [`d82ffca`](https://github.com/apache/spark/commit/d82ffca19c84589dfcb87a81996fb3ce2e31e91a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

gatorsmile commented:
cc @liufengdb

liufengdb commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -280,13 +280,20 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
   }
 
+  private[spark] def executeCollectIterator(): (Long, Iterator[InternalRow]) = {
+    val countsAndBytes = getByteArrayRdd().collect()
```
This still fetches all the compressed rows to the driver, before building the hashed relation. Ideally, you should fetch the rows from executors incrementally.

rdblue replied:
I agree, but this is better than before. This only keeps all of the rows in memory compressed, and then streams through the compressed blocks. Before this patch, the rows are copied into a buffer per row while holding the compressed blocks, so it held the rows compressed and uncompressed at the same time. The uncompressed rows are what this fixes, we can follow up with something better to stream through blocks from executors.

rdblue committed:
SPARK-22170: Fix SQL metrics test.

This includes a minor fix to an error message.

SparkQA commented:
**[Test build #82486 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82486/testReport)** for PR 19394 at commit [`3b43e11`](https://github.com/apache/spark/commit/3b43e112d11051c6ce88490530a3dbf4327efad3).

rdblue commented on sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:
```diff
@@ -228,7 +228,7 @@ class SQLMetricsSuite extends SparkFunSuite with SQLMetricsTestUtils with Shared
       1L -> (("BroadcastHashJoin", Map(
         "number of output rows" -> 2L,
-        "avg hash probe (min, med, max)" -> "\n(1, 1, 1)"))))
+        "avg hash probe (min, med, max)" -> "\n(1.1, 1.1, 1.1)"))))
```
This test was incorrect. I debugged the metric and the value is consistently 1.1, with 9 lookups and 10 probes. 10/9 = 1.1... The reason why the test was passing before is that the metrics check was hitting a [race condition](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsTestUtils.scala#L165-L169) that caused the test to pass. If I change the values in master to 1.1, the tests still pass and demonstrate that the check isn't really happening.

rdblue commented:
@rxin, any idea why this would fail ConfigBehaviorSuite? I don't think the failure is related because that test doesn't use a broadcast join. Should I rebase on master?

gatorsmile commented:
This is a pretty cool fix

gatorsmile commented:
Is the fix being used in the Netflix production environment?

rdblue commented:
Yes, we've been running this in production for a few weeks now.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:
```diff
@@ -73,25 +73,37 @@ case class BroadcastExchangeExec(
+            case arr: Array[InternalRow] =>
+              arr.map(_.asInstanceOf[UnsafeRow].getSizeInBytes.toLong).sum
+            case _ =>
+              numRows * 512 // guess: each row is about 512 bytes
```
We did the check in line 99. Thus, this could cause a false alarm. That means, this could cause a regression, right?

rdblue replied:
The won't cause a regression because all the broadcast modes return either Array or HashedRelation. This is just in case there is a path that returns something different in the future. Maybe it would be better to throw an exception here. What do you think?

gatorsmile replied:
Yeah, it makes more sense to issue an exception.

rdblue replied:
Fixed.

SparkQA commented:
**[Test build #82486 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82486/testReport)** for PR 19394 at commit [`3b43e11`](https://github.com/apache/spark/commit/3b43e112d11051c6ce88490530a3dbf4327efad3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82486/
Test FAILed.

rxin commented:
Not sure - maybe print the chi-value of the test and see if they make sense. If they do, we can change the threshold.

rdblue commented:
Here's the error message: TestFailedException: 347.5272 was not greater than 1000

rxin commented:
What's the other value?

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:
```diff
@@ -73,25 +73,37 @@ case class BroadcastExchangeExec(
         try {
           val beforeCollect = System.nanoTime()
           // Note that we use .executeCollect() because we don't want to convert data to Scala types
```
This comment is obsolete.

rdblue replied:
Updated.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:
```diff
@@ -73,25 +73,37 @@ case class BroadcastExchangeExec(
+
+          val dataSize = relation match {
+            case map: HashedRelation =>
+              map.estimatedSize
```
The estimated size of hashed relation maybe larger than the accurate data size calculated before. I guess this may be regression under some cases.

rdblue replied:
True, but I think using the accurate size is better. Maybe we should revisit the 8gb maximum instead of using a size that isn't correct. What's the reason for the 8gb max?

In master, building a broadcast table just under the 8gb limit would require 8gb for the rows (what we actually count here), plus the size of those rows compressed, which is probably at least 1gb. So we're talking about a pretty huge broadcast table to hit the case where this check will cause failures that didn't happen before. Not that it couldn't happen, but I've never seen this in practice and we have some huge jobs.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala:
```diff
@@ -73,25 +73,37 @@ case class BroadcastExchangeExec(
+          val dataSize = relation match {
+            case map: HashedRelation =>
+              map.estimatedSize
+            case arr: Array[InternalRow] =>
```
`mode` is a `BroadcastMode`. So I think we won't hit this case?

kiszk replied:
[`IdentityBroadcastMode.transform`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/broadcastMode.scala#L37) seems to return `Array[InternalRow]`.

viirya replied:
Yeah, right. `BroadcastNestedLoopJoinExec` uses `IdentityBroadcastMode`.

rdblue replied:
Yeah, both cases are used. The only one that won't match is `case _ => ...`

rdblue committed:
SPARK-22170: Fix ConfigBehaviorSuite, size logic.

SparkQA commented:
**[Test build #82524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82524/testReport)** for PR 19394 at commit [`f6613c2`](https://github.com/apache/spark/commit/f6613c2651691ca149390ea2407856ba390590cb).

rdblue commented on sql/core/src/test/scala/org/apache/spark/sql/ConfigBehaviorSuite.scala:
```diff
@@ -58,7 +58,7 @@ class ConfigBehaviorSuite extends QueryTest with SharedSQLContext {
         // If we only sample one point, the range boundaries will be pretty bad and the
         // chi-sq value would be very high.
-        assert(computeChiSquareTest() > 1000)
+        assert(computeChiSquareTest() > 300)
```
@rxin, the difference that is causing this to fail is that `rdd.id` is 15 instead of 14 because of the change to `getByteArrayRdd`. That id is used to seed the random generator and the chi-sq value isn't high enough. The previous value must have been unusually high. With so few data points, this can vary quite a bit so I think changing the bound to 300 is a good fix.

rdblue replied:
Updating the SparkPlan code to avoid creating unnecessary RDDs fixed this test because the same random seed is used. But, I think we should keep this change so we don't have to track down this problem again in the future. This bound is perfectly safe because the balanced chi-sq value is 10, while the unbalanced is much larger.

rdblue committed:
SPARK-22170: Fix comment.

SparkQA commented:
**[Test build #82525 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82525/testReport)** for PR 19394 at commit [`5b92fe2`](https://github.com/apache/spark/commit/5b92fe252530f4c8dab4d41772011e86ac308f80).

SparkQA commented:
**[Test build #82525 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82525/testReport)** for PR 19394 at commit [`5b92fe2`](https://github.com/apache/spark/commit/5b92fe252530f4c8dab4d41772011e86ac308f80).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82525/
Test FAILed.

SparkQA commented:
**[Test build #82524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82524/testReport)** for PR 19394 at commit [`f6613c2`](https://github.com/apache/spark/commit/f6613c2651691ca149390ea2407856ba390590cb).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82524/
Test FAILed.

rdblue commented:
Anyone have a clue what the python error could be? It doesn't look related.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #82530 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82530/testReport)** for PR 19394 at commit [`5b92fe2`](https://github.com/apache/spark/commit/5b92fe252530f4c8dab4d41772011e86ac308f80).

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -280,13 +280,20 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
+  private[spark] def executeCollectIterator(): (Long, Iterator[InternalRow]) = {
+    val countsAndBytes = getByteArrayRdd().collect()
+    val total = countsAndBytes.map(_._1).sum
+    val rows = countsAndBytes.iterator.map(_._2).flatMap(decodeUnsafeRows)
```
-> `flatMap(o => decodeUnsafeRows(o._2))`

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -280,13 +280,20 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    */
   def executeToIterator(): Iterator[InternalRow] = {
-    getByteArrayRdd().toLocalIterator.flatMap(decodeUnsafeRows)
+    getByteArrayRdd().toLocalIterator.map(_._2).flatMap(decodeUnsafeRows)
```
The same here.

rdblue replied:
Good point. I guess from the tests that this will create a new RDD, which isn't what we want.

rdblue replied:
Fixed.

rdblue committed:
SPARK-22170: Avoid creating unnecessary RDDs.

SparkQA commented:
**[Test build #82532 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82532/testReport)** for PR 19394 at commit [`56089f5`](https://github.com/apache/spark/commit/56089f5ba65f1d7d9e11b76673bcde3df37cd240).

SparkQA commented:
**[Test build #82530 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82530/testReport)** for PR 19394 at commit [`5b92fe2`](https://github.com/apache/spark/commit/5b92fe252530f4c8dab4d41772011e86ac308f80).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82530/
Test FAILed.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -274,19 +274,26 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    */
   def executeToIterator(): Iterator[InternalRow] = {
-    getByteArrayRdd().toLocalIterator.flatMap(decodeUnsafeRows)
+    getByteArrayRdd().toLocalIterator.flatMap(rdd => decodeUnsafeRows(rdd._2))
```
We don't need to collect the counts back to driver. Besides it should not be a rdd but bytes in the `flatMap`.

Maybe:

```scala
getByteArrayRdd().map(_._2).toLocalIterator.flatMap(decodeUnsafeRows(_))
```

rdblue replied:
Fixed.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -274,19 +274,26 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
+  private[spark] def executeCollectIterator(): (Long, Iterator[InternalRow]) = {
+    val countsAndBytes = getByteArrayRdd().collect()
+    val total = countsAndBytes.map(_._1).sum
+    val rows = countsAndBytes.iterator.flatMap(rdd => decodeUnsafeRows(rdd._2))
```
nit: `rdd` maybe confusing.

rdblue replied:
Fixed.

SparkQA commented:
**[Test build #82532 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82532/testReport)** for PR 19394 at commit [`56089f5`](https://github.com/apache/spark/commit/56089f5ba65f1d7d9e11b76673bcde3df37cd240).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82532/
Test PASSed.

rdblue committed:
SPARK-22170: Minor changes for review comments.

rdblue committed:
SPARK-22170: Fix naming.

SparkQA commented:
**[Test build #82555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82555/testReport)** for PR 19394 at commit [`a2976fe`](https://github.com/apache/spark/commit/a2976fe093f0b0afe9a8b01a876d401fb0bf2d04).

SparkQA commented:
**[Test build #82556 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82556/testReport)** for PR 19394 at commit [`1a813ac`](https://github.com/apache/spark/commit/1a813acecb61933544c7c6ca1ebd58d5b95b7fa5).

gatorsmile commented:
LGTM

SparkQA commented:
**[Test build #82556 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82556/testReport)** for PR 19394 at commit [`1a813ac`](https://github.com/apache/spark/commit/1a813acecb61933544c7c6ca1ebd58d5b95b7fa5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82556/
Test FAILed.

rdblue committed:
SPARK-22170: Fix flaky SQL metrics test.

The "BroadcastHashJoin metrics" test is flaky. Previously, the metrics
weren't captured or checked because of the race condition mentioned in
SQLMetricsTestUtils.getSparkPlanMetrics. After the memory fix, the
average hash probe metric is not reliable. Sometimes it is 1.1 and
sometimes 1.2. This should be fixed when the race condition in the
metrics test utility is fixed.

SparkQA commented:
**[Test build #82555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82555/testReport)** for PR 19394 at commit [`a2976fe`](https://github.com/apache/spark/commit/a2976fe093f0b0afe9a8b01a876d401fb0bf2d04).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82555/
Test PASSed.

SparkQA commented:
**[Test build #82558 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82558/testReport)** for PR 19394 at commit [`ab2e3ca`](https://github.com/apache/spark/commit/ab2e3cabebc7584e3803cffc727aad5b8cd16122).

SparkQA commented:
**[Test build #82558 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82558/testReport)** for PR 19394 at commit [`ab2e3ca`](https://github.com/apache/spark/commit/ab2e3cabebc7584e3803cffc727aad5b8cd16122).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82558/
Test PASSed.

gatorsmile commented:
Thanks! Merged to master.

rdblue commented:
Thanks for reviewing, @gatorsmile!
